### ISSUE TITLE
Nitrokey Pro 0.9 -> 0.14 impossible

### ISSUE THREAD

**Issue by tlaurion** [Nitrokey/nitrokey-pro-firmware#95]
Hello there.

A user reported a Nitrokey Pro delivery delayed of 6 weeks. He decided to not trust received dongle and reflash it with newer firmware thrhough dfu external programmer, which at the time was 0.14 RC4 which had GPIO mapping inversed, which led to issue https://github.com/osresearch/heads/issues/1170

The user bought programmer and went the manual way.
- I was not aware, either, of a software tool that permitted to do the flashing without external programmer, which is documented deep down under https://github.com/Nitrokey/nitrokey-pro-firmware/blob/master/DEVELOPMENT.md#firmware-update

I was able to upgrade with that tool from Nitrokey's firmware:
- 0.10
- 0.11

But for some reason, attempting to upgrade 0.9 was not possible.
I got firware version from heads through `hotp_verification info`

0.10 and 0.11 firmware version behaves the following way:
- `nitropy pro enable-update` triggers an internal reboot of the dongle, and the idea is different on next boot in update mode.
- `nitropy pro update nitrokey-pro-firmware-v0.14-to_update.bin` is successful.

0n 0.9 connected dongle:
- Qubes dom0 blocks keyboard (unrelevant) on usb dongle connection
- `nitropy pro enable-update` doesn't trigger a reboot of the dongle and device stays attached to affected to qubes where tools are installed.
- `nitropy pro update nitrokey-pro-firmware-v0.14-to_update.bin` cannot find dongle in update mode.

Questions needing answers:
- Are 0.9 firmware based dongle non-updateable with nitropy?
- Librem Key are not having proper udev rules defined and are not upgradeable through nitropy. Is that expected? @kylerankin @jans23 @MrChromebox?
- Wouldn't we want users to be more aware of the possibility of upgrading their dongles since it is now possible with software, and document limitations?

-------

Some unrelated issues here:
- Searching Nitrokey website for Nitrokey Pro led flashing patterns doesn't give info as before. Section removed from FAQ?

**Comment by szszszsz** [comment]
Hi!
1. The firmware update feature is available only since v0.11, meaning the Nitrokey Pro v0.10 devices and older cannot be updated by solely software means as well. Additionally one can see that the size of the binary increases between these two by about 20 kB, which is taken by the bootloader code.
- https://github.com/Nitrokey/nitrokey-pro-firmware/releases/tag/v0.11
2. About this specific issue 1170, the firmware reported in the very first screenshot was v0.13, however v0.14RC4 should contain the updated firmware string already (as specified in the repository), hence I suspect the user was really using v0.14RC1 at the time (if not v0.13) and not having the correct LED behavior yet. Perhaps some misunderstanding?
3. We are finishing the work on the firmware update tool, specifically MS Windows support, and plan to publish it.
4. As far as I remember Librem Keys have own Udev rules, with own distribution path. Naturally they might not contain update mode IDs yet, since the tool was not considered to be ready.
5. I have passed further the docs check request.

**Comment by tlaurion** [comment]
Thanks! You're right, only 0.11+ were upgraded, 0.10 device is a Librem Key. Edited OP.

Right as well for user report, was 0.13.

I would correct flashing references to clarify this as well. 0.11 software upgradeable would help redirection support. Also note that RPI can be used as flasher and should also be referenced (why have multiple flashers if RPI can be used and is handy.

**Comment by tlaurion** [comment]
Note on current instructions.

Debian-11 based appvm under Qubes OS.
- Having nitrokey-app installed on debian-11 fulfills dependencies for libnitrokey
- `sudo apt install python3-pip` can be installed in volatile qube (will not persist reboot)
- `pip3 install --user pynitrokey` installs pynitrokey in user (private volume) and persists reboot.
- Pass dongle to qube
- `nitropy pro enable-update`, disconnects device from qube (not same ID).
- Pass dongle to qube
- `nitropy pro update nitrokey-pro-firmware-v0.14-to_update.bin` not successful, complaining "Cannot connect to the device: Bus 001 Device 003: ID 20a0:42b4"
- Have to download and apply udev referred rules on command line with `curl https://raw.githubusercontent.com/Nitrokey/libnitrokey/master/data/41-nitrokey.rules | sudo tee /usr/lib/udev/rules.d/41-nitrokey.rules`
- Unfortunately, calling `sudo udevadm control --reload-rules; sudo udevadm trigger` as said on console from `nitropy pro update` was not successful, and would require the udev rules download, reload and trigger calls to be followed by a prompt asking the user to physically unplug and replug dongle and redo the steps above from "Pass dongle to qube" to "nitropy pro update nitrokey-pro-firmware-v0.14-to_update.bin".
- Sucess

@szszszsz that was my successful experience upgrading 0.11 dongles to 0.14, hope this finds a way into the documentation somehow.

**Comment by tlaurion** [comment]
@szszszsz 

Hello there,

Documentation specifies 

"This approach requires soldering wires to the contact pads or to use an adapter with pogo pins and some kind of mounting (recommended)."

Any picture or more specifics to share?

**Comment by szszszsz** [comment]
@tlaurion Hi!
I do not think we ended up with the proper mounting for the development, but will pass the question further and get back if there would be anything of use.
We do have pogo pins on a header for the production purposes, though its not public. You can see it in action here:
- https://youtu.be/kqUK9cAjH44?t=32

About the actual soldering via contact pads for the SWD connection, the details are just below:
- https://github.com/Nitrokey/nitrokey-pro-firmware#swd

**Comment by szszszsz** [comment]
@tlaurion Attaching image of the pogo pin adapter below (centered on the actual pins connecting to the PCB):
![image](https://user-images.githubusercontent.com/17005426/176458815-d38705b0-8749-4932-a710-2e409c2a055e.png)

**Comment by aanno** [comment]
Perhaps a few years late, but same problem here. However, I'm not thrilled by soldering or using openocd. But I think the [flashing](https://github.com/Nitrokey/nitrokey-pro-firmware#flashing) instruction is incomplete. ocd script is for backing up old firmware. I guess writing the firmware is just about the same (but with `write_bank`), right?

BTW, is there any way to find out the current firmware version?

**Comment by tlaurion** [comment]
> BTW, is there any way to find out the current firmware version?

Not Nitrokey dev here, but nitropy should give versioning output.

Note: Iwas pinged because I opened the issue, changed name of issue at the same time of answering here. 

Tldr:external flashing needed <= 0.10

**Comment by aanno** [comment]
Well, 

I tried nitropy. However, it does _not_ give me the firmware version of my device. But it says 'The connected Nitrokey Pro does not support firmware updates'. Does that mean that using `openocd` for firmware update is an option?

```sh
$ nitropy pro 
Command line tool to interact with Nitrokey devices 0.12.0
Usage: nitropy pro [OPTIONS] COMMAND [ARGS]...

  Interact with Nitrokey Pro devices, see subcommands.

Options:
  -h, --help  Show this message and exit.

Commands:
  change-firmware-password  Change the firmware update password.
  enable-update             enable firmware update for NK Pro device
  list                      list connected devices
  update                    Run firmware update with the provided binary.

$ nitropy pro list
Command line tool to interact with Nitrokey devices 0.12.0
:: 'Nitrokey Pro' keys:
Nitrokey Pro-39CF 1-3.1.1:1.0

$ nitropy pro change-firmware-password
Command line tool to interact with Nitrokey devices 0.12.0
Critical error:
The connected Nitrokey Pro does not support firmware updates
```

I also tried to update the firmware, anyway. Please note that I have libnitrokey _installed_ and nitrokey-app running.

```sh
$ ls -l /usr/lib64/libnitrokey.so*
lrwxrwxrwx. 1 root root      20 16. Jan 01:00 /usr/lib64/libnitrokey.so.3 -> libnitrokey.so.3.7.0
-rwxr-xr-x. 1 root root 1433920 16. Jan 01:00 /usr/lib64/libnitrokey.so.3.7.0

$ ls -la firmware/
insgesamt 372
drwxr-xr-x.  2 tpasch tpasch   4096 25. Mai 2022  .
drwxr-xr-x. 12 tpasch tpasch   4096 11. Apr 17:33 ..
-rw-r--r--.  1 tpasch tpasch   4300 25. Mai 2022  nitrokey-pro-firmware.buildinfo
-rw-r--r--.  1 tpasch tpasch 253332 25. Mai 2022  nitrokey-pro-firmware-v0.15-to_flash.hex
-rw-r--r--.  1 tpasch tpasch    171 25. Mai 2022  nitrokey-pro-firmware-v0.15-to_flash.hex.sha512
-rwxr-xr-x.  1 tpasch tpasch  98760 25. Mai 2022  nitrokey-pro-firmware-v0.15-to_update.bin
-rw-r--r--.  1 tpasch tpasch    172 25. Mai 2022  nitrokey-pro-firmware-v0.15-to_update.bin.sha512

$ nitropy pro update firmware/nitrokey-pro-firmware-v0.15-to_update.bin
Command line tool to interact with Nitrokey devices 0.12.0
No Nitrokey Pro found in the update mode.
If you have Nitrokey Pro connected please run (requires libnitrokey):
  $ nitropy pro enable-update

$ nitropy pro enable-update
Command line tool to interact with Nitrokey devices 0.12.0
Enabling firmware update mode
No Nitrokey Pro device found
If connected, perhaps already in update mode?

# still the same afterwards
$ nitropy pro update firmware/nitrokey-pro-firmware-v0.15-to_update.bin
...
```

**Comment by jans23** [comment]
Depending on the hardware revision, you would need to open the device, connect two pins to start the boot loader which would allow you to flash with tools such stm32flash.
